src/btree/bt_cursor.c

```diff
@@ -512,12 +512,40 @@ __wt_btcur_search_prepared(WT_CURSOR *cursor, WT_UPDATE **updp)
     return (0);
 }
 
+/*
+ * __wt_btcur_evict_reposition --
+ *     Try to evict the page and reposition the cursor on the saved key.
+ */
+int
+__wt_btcur_evict_reposition(WT_CURSOR_BTREE *cbt)
+{
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+
+    cursor = &cbt->iface;
+    session = CUR2S(cbt);
+
+    if (session->txn->isolation == WT_ISO_SNAPSHOT && !WT_SESSION_IS_CHECKPOINT(session) &&
+      F_ISSET_ATOMIC_16(cbt->ref->page, WT_PAGE_FORCE_EVICTION)) {
+        F_SET(cbt, WT_CBT_RELEASE_EVICT_PAGE);
```

Comment:
I wonder if we could avoid setting this flag (WT_PAGE_FORCE_EVICTION) and passing parameters around to cursor operations. It's really the page state that matters, there is a check here:
https://github.com/wiredtiger/wiredtiger/blame/develop/src/include/btree_inline.h#L1799
That I think should do what we are looking for, i.e: checks to see whether the page requires urgent eviction and if it is possible that this thread can help with that.

I wonder if we can pull the first part of the conditional check out into a separate function, something like:

```
bool __wt_page_evict_soon_check(WT_SESSION_IMPL *session, WT_REF *ref)
{
    if (WT_READGEN_EVICT_SOON(page->read_gen) && btree->evict_disabled == 0 &&
      __wt_page_can_evict(session, ref, &inmem_split) &&
      (!WT_SESSION_IS_CHECKPOINT(session) || __wt_page_evict_clean(page)))
        return (true);
    return (false);
}
```

If we use that condition, we also shouldn't need to set any special flags on the page or cursor. The cursor reset should lead to `__wt_page_release` being called, which will result in the page being either evicted or queued for urgent eviction.

Reply:
I think this makes sense to me. The only concern is that __wt_page_can_evict(session, ref, &inmem_split) may be costly and we need to pass in the inmem_split flag.

Reply:
I agree, this seems like a better approach.

Reply:
Updated in [d63e6ca](https://github.com/wiredtiger/wiredtiger/pull/7687/commits/d63e6ca7853441e2d2ca278f0a9ae4c8a81125f7)